simulator/simulator.go

```diff
@@ -0,0 +1,329 @@
+package simulator
+
+import (
+	"context"
+	"crypto/tls"
+	"net"
+	"sync"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/go-kit/kit/log"
+	"github.com/kolide/launcher/service"
+	"github.com/kolide/osquery-go/plugin/distributed"
+	"github.com/kolide/osquery-go/plugin/logger"
+	"github.com/pkg/errors"
+)
+
+// QueryRunner is the interface which defines the pluggable behavior of a simulated
+// host. Each "type" of host may have their own implementation of this interface.
+type QueryRunner interface {
+	RunQuery(sql string) (results []map[string]string, err error)
+}
+
+// HostSimulation is the type which contains the state of a simulated host
+type HostSimulation struct {
+	// the following define the configurable aspects of the simulation
+	host                   QueryRunner
+	uuid                   string
+	enrollSecret           string
+	grpcURL                string
+	insecure               bool
+	insecureGrpc           bool
+	requestQueriesInterval time.Duration
+	requestConfigInterval  time.Duration
+	publishLogsInterval    time.Duration
+
+	// The state of the simulation is gated with a read/write lock.
+	// To read something in state:
+	//
+	//   h.state.lock.RLock()
+	//   defer h.state.lock.RUnlock()
+	//
+	// To write state based on the on-going simulation:
+	//
+	//   h.state.lock.Lock()
+	//   defer h.state.lock.Unlock()
+	state *hostSimulationState
+
+	shutdown chan bool
```

Comment:
see my previous comment about using a single `chan chan struct{}` for the shutdown method.

Reply:
I don't know, I am not super convinced that a channel of channels is cleaner than two channels really. Do you have any further reading on this topic?

Reply:
No reading, just examples 
https://github.com/oklog/oklog/blob/d78d4a9b54097c70a3abb4a793dca5f30cf9f4b4/pkg/ingest/api.go#L107-L111

There's a larger pattern within that code that has to do with creating actors which accept functions on a channel too that would be nice to discuss separately. I have some reading on the actor pattern specifically, not on the `stop chan chan struct{}` primitive. 

https://github.com/oklog/oklog/blob/d78d4a9b54097c70a3abb4a793dca5f30cf9f4b4/pkg/ingest/api.go#L32